File: apps/mobile/lib/domain/models/round.dart
// Round Model — VSP Mobile App
//
// Local round entity for offline-first round persistence.
// Stored in vsp_round.db.
//
// Story 5.2: Persist Round Locally
// Story 12.1 Slice F: tournamentId, tournamentPolicyVersion, tournamentPolicyJson

import 'dart:convert';
import 'package:equatable/equatable.dart';

import 'tournament_policy.dart';

/// Round lifecycle status.
enum RoundStatus { inProgress, completed, abandoned, cancelled }

/// Round entity — represents one golf round.
class Round extends Equatable {
  final String id; // UUID, local identifier
  final int courseId;

  /// The second đường, where the round pairs two nines.
  ///
  /// Long Biên has three of them and a round there is a pairing chosen on the
  /// day, so round hole 10 is hole 1 of whichever nine came second. The round
  /// setup screen has always known this and passed it into the round; the
  /// round itself did not carry it, so resuming one left holes 10 to 18
  /// belonging to no course. The map asked the front nine for its hole 10,
  /// which does not exist, and said the hole had not been surveyed.
  ///
  /// Null for an eighteen played on one course, which is most of them.
  final int? backNineCourseId;
  final String courseName; // denormalized from course package
  final RoundStatus status;
  final DateTime startedAt;
  final DateTime? endedAt;
  final String packageVersion; // captured at round start for guard
  final String? tournamentPolicyId; // null for casual/practice rounds
  /// Tournament ID — links this round to a tournament event.
  /// Per Story 12.1 Slice F.
  final String? tournamentId;

  /// Version of the tournament policy at round creation time.
  /// Used to detect mid-round policy changes.
  /// Per Story 12.1 Slice F.
  final int? tournamentPolicyVersion;

  /// Cached tournament policy JSON for offline restricted feature checks.
  /// Per Story 12.1 Slice F.
  final TournamentPolicy? tournamentPolicy;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Round({
    required this.id,
    required this.courseId,
    this.backNineCourseId,
    required this.courseName,
    required this.status,
    required this.startedAt,
    this.endedAt,
    required this.packageVersion,
    this.tournamentPolicyId,
    this.tournamentId,
    this.tournamentPolicyVersion,
    this.tournamentPolicy,
    required this.createdAt,
    required this.updatedAt,
  });

  /// True if this round is still in progress.
  bool get isActive => status == RoundStatus.inProgress;

  /// True if this is a tournament round (has a tournament policy).
  bool get isTournamentRound => tournamentPolicyId != null;

  /// Copy with updated fields.
  Round copyWith({
    String? id,
    int? courseId,
    int? backNineCourseId,
    String? courseName,
    RoundStatus? status,
    DateTime? startedAt,
    DateTime? endedAt,
    String? packageVersion,
    String? tournamentPolicyId,
    String? tournamentId,
    int? tournamentPolicyVersion,
    TournamentPolicy? tournamentPolicy,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Round(
      id: id ?? this.id,
      courseId: courseId ?? this.courseId,
      backNineCourseId: backNineCourseId ?? this.backNineCourseId,
      courseName: courseName ?? this.courseName,
      status: status ?? this.status,
      startedAt: startedAt ?? this.startedAt,
      endedAt: endedAt ?? this.endedAt,
      packageVersion: packageVersion ?? this.packageVersion,
      tournamentPolicyId: tournamentPolicyId ?? this.tournamentPolicyId,
      tournamentId: tournamentId ?? this.tournamentId,
      tournamentPolicyVersion:
          tournamentPolicyVersion ?? this.tournamentPolicyVersion,
      tournamentPolicy: tournamentPolicy ?? this.tournamentPolicy,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  /// Convert to a Map for SQLite persistence.
  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'course_id': courseId,
      'back_nine_course_id': backNineCourseId,
      'course_name': courseName,
      'status': status.name,
      'started_at': startedAt.toUtc().toIso8601String(),
      'ended_at': endedAt?.toUtc().toIso8601String(),
      'package_version': packageVersion,
      'tournament_policy_id': tournamentPolicyId,
      'tournament_id': tournamentId,
      'tournament_policy_version': tournamentPolicyVersion,
      'tournament_policy_json': tournamentPolicy != null
          ? jsonEncode(tournamentPolicy!.toJson())
          : null,
      'created_at': createdAt.toUtc().toIso8601String(),
      'updated_at': updatedAt.toUtc().toIso8601String(),
    };
  }

  /// Reconstruct from a SQLite row.
  factory Round.fromMap(Map<String, dynamic> map) {
    TournamentPolicy? policy;
    if (map['tournament_policy_json'] != null) {
      try {
        policy = TournamentPolicy.fromJson(
          jsonDecode(map['tournament_policy_json'] as String)
              as Map<String, dynamic>,
        );
      } catch (_) {
        policy = null;
      }
    }
    return Round(
      id: map['id'] as String,
      courseId: (map['course_id'] as num).toInt(),
      // Absent on a row written before this column existed, which every
      // in-progress round on an installed app is.
      backNineCourseId: (map['back_nine_course_id'] as num?)?.toInt(),
      courseName: map['course_name'] as String,
      status: RoundStatus.values.firstWhere(
        (e) => e.name == map['status'],
        orElse: () => RoundStatus.inProgress,
      ),
      startedAt: DateTime.parse(map['started_at'] as String),
      endedAt: map['ended_at'] != null
          ? DateTime.parse(map['ended_at'] as String)
          : null,
      packageVersion: map['package_version'] as String,
      tournamentPolicyId: map['tournament_policy_id'] as String?,
      tournamentId: map['tournament_id'] as String?,
      tournamentPolicyVersion: (map['tournament_policy_version'] as num?)
          ?.toInt(),
      tournamentPolicy: policy,
      createdAt: DateTime.parse(map['created_at'] as String),
      updatedAt: DateTime.parse(map['updated_at'] as String),
    );
  }

  @override
  List<Object?> get props => [
    id,
    courseId,
    backNineCourseId,
    courseName,
    status,
    startedAt,
    endedAt,
    packageVersion,
    tournamentPolicyId,
    tournamentId,
    tournamentPolicyVersion,
    tournamentPolicy,
    createdAt,
    updatedAt,
  ];
}

File: apps/api/src/main/java/vnpt/vsp/module/round/entity/Round.java
package vnpt.vsp.module.round.entity;

import jakarta.persistence.*;
import java.time.Instant;
import java.util.UUID;

/**
 * Round entity representing a golf round.
 * Per PRD: supports in_progress, completed, abandoned, cancelled statuses.
 */
@Entity
@Table(name = "rounds")
public class Round {

    public enum RoundStatus {
        IN_PROGRESS,
        COMPLETED,
        ABANDONED,
        CANCELLED
    }

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    /**
     * The course this round is played on.
     * TODO ( Slice D gap): V13 migration does not include course_id column.
     * A follow-up migration V14 should add: ALTER TABLE rounds ADD COLUMN course_id BIGINT REFERENCES courses(id)
     */
    @Column(name = "course_id")
    private Long courseId;

    /**
     * The second nine, where the round is composed of two.
     *
     * <p>Long Biên is đường A, B and C at nine holes each, so an eighteen there
     * is two of them. A round could name one course, which left holes 10 to 18
     * belonging to nothing: the map had no survey for them, advice answered
     * 404, and the pars and yardages this project had already loaded were
     * unreachable.
     *
     * <p>Null where the round is one eighteen-hole layout, which is most of
     * them — {@link #courseId} then means the whole course, exactly as before.
     */
    @Column(name = "back_nine_course_id")
    private Long backNineCourseId;

    @Column(name = "golfer_account_id", nullable = false)
    private Long golferAccountId;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 20, nullable = false)
    private RoundStatus status = RoundStatus.IN_PROGRESS;

    @Column(name = "started_at", nullable = false)
    private Instant startedAt;

    @Column(name = "ended_at")
    private Instant endedAt;

    @Column(name = "deleted_at")
    private Instant deletedAt;

    /**
     * Tournament policy ID for tournament-format rounds.
     * Required when format is TOURNAMENT; null for casual and practice rounds.
     */
    @Column(name = "tournament_policy_id")
    private UUID tournamentPolicyId;

    /**
     * Tournament ID — links the round to a tournament event.
     * When set, the round is a tournament round and submits scores to the tournament.
     * Per Story 12.1 Slice F.
     */
    @Column(name = "tournament_id")
    private UUID tournamentId;

    /**
     * Version of the tournament policy at round creation time.
     * Used to detect mid-round policy changes — mobile shows update banner
     * when server policy version differs from this value.
     * Per Story 12.1 Slice F.
     */
    @Column(name = "tournament_policy_version")
    private Integer tournamentPolicyVersion;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = Instant.now();
        updatedAt = Instant.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = Instant.now();
    }

    // ─── Getters and Setters ───────────────────────────────────────────────

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public Long getCourseId() {
        return courseId;
    }

    public void setCourseId(Long courseId) {
        this.courseId = courseId;
    }

    public Long getBackNineCourseId() {
        return backNineCourseId;
    }

    public void setBackNineCourseId(Long backNineCourseId) {
        this.backNineCourseId = backNineCourseId;
    }

    /**
     * Which course a hole of this round belongs to, and its number there.
     *
     * <p>Hole 10 of a round composed of two nines is hole 1 of the back nine.
     * Everything that asks about a hole — the map, the scorecard, the advice —
     * has to ask the course that actually holds it, and this is the one place
     * that translation lives.
     *
     * @return [courseId, holeNumberOnThatCourse]
     */
    public long[] resolveHole(int holeNumber) {
        if (backNineCourseId == null || holeNumber <= 9) {
            return new long[]{courseId == null ? 0L : courseId, holeNumber};
        }
        return new long[]{backNineCourseId, holeNumber - 9};
    }

    public Long getGolferAccountId() {
        return golferAccountId;
    }

    public void setGolferAccountId(Long golferAccountId) {
        this.golferAccountId = golferAccountId;
    }

    public RoundStatus getStatus() {
        return status;
    }

    public void setStatus(RoundStatus status) {
        this.status = status;
    }

    public Instant getStartedAt() {
        return startedAt;
    }

    public void setStartedAt(Instant startedAt) {
        this.startedAt = startedAt;
    }

    public Instant getEndedAt() {
        return endedAt;
    }

    public void setEndedAt(Instant endedAt) {
        this.endedAt = endedAt;
    }

    public Instant getDeletedAt() {
        return deletedAt;
    }

    public void setDeletedAt(Instant deletedAt) {
        this.deletedAt = deletedAt;
    }

    public UUID getTournamentPolicyId() { return tournamentPolicyId; }
    public void setTournamentPolicyId(UUID tournamentPolicyId) { this.tournamentPolicyId = tournamentPolicyId; }

    public UUID getTournamentId() { return tournamentId; }
    public void setTournamentId(UUID tournamentId) { this.tournamentId = tournamentId; }

    public Integer getTournamentPolicyVersion() { return tournamentPolicyVersion; }
    public void setTournamentPolicyVersion(Integer tournamentPolicyVersion) { this.tournamentPolicyVersion = tournamentPolicyVersion; }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }
}
